File: ruby/lib/redis/commands/get.rb
require_relative './base'

module Redis
  module Commands
    class Get < Base
      def self.command
        'get'
      end
      
      def run
        @mutex.synchronize do 
          key = args[0]
          value = @store[key]
          return Resp::Serializer.serialize(nil) unless value
          
          # Passive expire
          if key_expired?(key)
            @store.delete(key)
            Resp::Serializer.serialize(nil)
          else
            Resp::Serializer.serialize(value)
          end
        end
      end

      private

      def key_expired?(key)
        # key = key.to_sym
        if @expire_keys[key] && @expire_keys[key] < Time.now.to_i * 1000
          @expire_keys.delete(key)
          return true
        end
  
        false
      end
    end
  end
end

File: python/src/commands/get.py
from .base import BaseCommand
from src.resp.serializer import RespSerializer
from time import time

class Get(BaseCommand):
  @classmethod
  def command(cls) -> bytes:
    return b'get'
  
  def run(self) -> bytes:
    key: bytes = self.args[0]
    if self._has_key_expired(key=key):
      self.expire_store.pop(key)
      self.store.pop(key)
      return RespSerializer.serialize(data=None)

    return RespSerializer.serialize(data=self.store.get(key))

  def _has_key_expired(self, key: bytes) -> bool:
    ttl = self.expire_store.get(key)
    if ttl:
      return True if time() * 1000 > ttl else False

    return False
  